scene/3d/xr_hand_modifier_3d.cpp

```diff
@@ -0,0 +1,309 @@
+void XRHandModifier3D::set_bone_update(BoneUpdate p_bone_update) {
+	ERR_FAIL_INDEX(p_bone_update, BONE_UPDATE_MAX);
+	bone_update = p_bone_update;
+}
+
+XRHandModifier3D::BoneUpdate XRHandModifier3D::get_bone_update() const {
+	return bone_update;
+}
+
+Skeleton3D *XRHandModifier3D::get_skeleton() {
+	if (!has_node(target)) {
+		return nullptr;
+	}
+
+	Node *node = get_node(target);
+	if (!node) {
+		return nullptr;
+	}
+
+	Skeleton3D *skeleton = Object::cast_to<Skeleton3D>(node);
+	return skeleton;
+}
+
+void XRHandModifier3D::_get_joint_data() {
+	// Table of bone names for different rig types.
+	static const String bone_names[XRHandTracker::HAND_JOINT_MAX] = {
+		"Palm",
+		"Hand",
+		"ThumbMetacarpal",
+		"ThumbProximal",
+		"ThumbDistal",
+		"ThumbTip",
+		"IndexMetacarpal",
+		"IndexProximal",
+		"IndexIntermediate",
+		"IndexDistal",
+		"IndexTip",
+		"MiddleMetacarpal",
+		"MiddleProximal",
+		"MiddleIntermediate",
+		"MiddleDistal",
+		"MiddleTip",
+		"RingMetacarpal",
+		"RingProximal",
+		"RingIntermediate",
+		"RingDistal",
+		"RingTip",
+		"LittleMetacarpal",
+		"LittleProximal",
+		"LittleIntermediate",
+		"LittleDistal",
+		"LittleTip",
+	};
+
+	static const String bone_name_format[2] = {
+		"Left<bone>",
+		"Right<bone>",
+	};
+
+	// reset JIC
+	for (int i = 0; i < XRHandTracker::HAND_JOINT_MAX; i++) {
+		joints[i].bone = -1;
+		joints[i].parent_joint = -1;
+	}
+
+	Skeleton3D *skeleton = get_skeleton();
+	if (!skeleton) {
+		return;
+	}
+
+	XRServer *xr_server = XRServer::get_singleton();
+	if (!xr_server) {
+		return;
+	}
+
+	Ref<XRHandTracker> tracker = XRServer::get_singleton()->get_hand_tracker(tracker_name);
+	if (tracker.is_null()) {
+		return;
+	}
+
+	XRHandTracker::Hand hand = tracker->get_hand();
+
+	// Find the skeleton-bones associated with each joint.
+	int bones[XRHandTracker::HAND_JOINT_MAX];
+	for (int i = 0; i < XRHandTracker::HAND_JOINT_MAX; i++) {
+		// Construct the expected bone name.
+		String bone_name = bone_name_format[hand].replace("<bone>", bone_names[i]);
+
+		// Find the skeleton bone.
+		bones[i] = skeleton->find_bone(bone_name);
+		if (bones[i] == -1) {
+			print_line("Couldn't obtain bone for", bone_name);
+		}
+	}
+
+	// Assemble the joint relationship to the available skeleton bones.
+	for (int i = 0; i < XRHandTracker::HAND_JOINT_MAX; i++) {
+		// Get the skeleton bone (skip if not found).
+		const int bone = bones[i];
+		if (bone == -1) {
+			continue;
+		}
+
+		// Find the parent skeleton-bone.
+		const int parent_bone = skeleton->get_bone_parent(bone);
+		if (parent_bone == -1) {
+			// If no parent skeleton-bone exists then drive this relative to palm joint.
+			joints[i].bone = bone;
+			joints[i].parent_joint = XRHandTracker::HAND_JOINT_PALM;
+			continue;
+		}
+
+		// Find the joint associated with the parent skeleton-bone.
+		for (int j = 0; j < XRHandTracker::HAND_JOINT_MAX; ++j) {
+			if (bones[j] == parent_bone) {
+				// If a parent joint is found then drive this bone relative to it.
```

Comment:
If the parent joint is not found, should it match the above logic and make it relative to the palm joint?

Reply:
Hm. @Malcolmnixon would know better since I think this is his code originally, but I think if we have a parent bone that isn't related to any valid hand joint from the Godot Humanoid skeleton, we don't want to drive it at all - which is what I think would happen with the current code in the PR.

I think the earlier case where we're checking for `parent_bone == -1` is about handling the root of the skeleton, which could start on either a palm or wrist joint.